server/src/main/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinator.java

```diff
@@ -258,9 +258,73 @@ public int markSegmentsAsUnusedWithinInterval(String dataSource, Interval interv
 
   /**
    * Fetches all the pending segments, whose interval overlaps with the given
-   * search interval from the metadata store. Returns a Map from the
-   * pending segment ID to the sequence name.
+   * search interval and has a sequence_name that begins with one of the prefixes in sequenceNamePrefixFilter
+   * from the metadata store. Returns a Map from the pending segment ID to the sequence name.
    */
+  @VisibleForTesting
+  Map<SegmentIdWithShardSpec, String> getPendingSegmentsForIntervalWithHandle(
+      final Handle handle,
+      final String dataSource,
+      final Interval interval,
+      final Set<String> sequenceNamePrefixFilter
+  ) throws IOException
+  {
+    if (sequenceNamePrefixFilter.isEmpty()) {
+      return Collections.emptyMap();
+    }
+
+    final List<String> sequenceNamePrefixes = new ArrayList<>(sequenceNamePrefixFilter);
+    StringBuilder sql = new StringBuilder(
+        "SELECT sequence_name, payload FROM "
+        + dbTables.getPendingSegmentsTable()
+        + " WHERE dataSource = :dataSource AND start < :end AND "
+        + connector.getQuoteString() + "end" + connector.getQuoteString() + " > :start"
+    );
+
+    sql.append(" AND ( ");
+
+    for (int i = 1; i < sequenceNamePrefixes.size(); i++) {
+      sql.append("(sequence_name LIKE ")
+         .append(StringUtils.format(":prefix%d", i))
+         .append(")")
+         .append(" OR ");
+    }
+
+    sql.append("(sequence_name LIKE ")
+       .append(StringUtils.format(":prefix%d", sequenceNamePrefixes.size()))
+       .append(")");
+
+    sql.append(" )");
+
+    Query<Map<String, Object>> query =
+        handle.createQuery(sql.toString())
+              .bind("dataSource", dataSource)
+              .bind("start", interval.getStart().toString())
+              .bind("end", interval.getEnd().toString());
+
+    for (int i = 1; i <= sequenceNamePrefixes.size(); i++) {
+      query.bind(StringUtils.format("prefix%d", i), sequenceNamePrefixes.get(i - 1) + "%");
```

Comment:
What do we need the `%` for in `sequenceNamePrefixes.get(i - 1) + "%"`

Reply:
Because we want the sequence to begin with the sequence name prefix and not just be equal to it